plugins/hlint-hls-plugin/src/Ide/Plugin/Hlint.hs

```diff
@@ -0,0 +1,701 @@
+        [(file, ShowDiag, ideaToDiagnostic i) | i <- ideas, ideaSeverity i /= Ignore]
+      diagnostics file (Left parseErr) =
+        [(file, ShowDiag, parseErrorToDiagnostic parseErr)]
+
+      ideaToDiagnostic :: Idea -> Diagnostic
+      ideaToDiagnostic idea =
+        LSP.Diagnostic {
+            _range    = srcSpanToRange $ ideaSpan idea
+          , _severity = Just LSP.DsInfo
+          , _code     = Just (LSP.StringValue $ T.pack $ ideaHint idea)
+          , _source   = Just "hlint"
+          , _message  = T.pack $ show idea
+          , _relatedInformation = Nothing
+          , _tags     = Nothing
+        }
+
+      parseErrorToDiagnostic :: ParseError -> Diagnostic
+      parseErrorToDiagnostic (Hlint.ParseError l msg contents) =
+        LSP.Diagnostic {
+            _range    = srcSpanToRange l
+          , _severity = Just LSP.DsInfo
+          , _code     = Just (LSP.StringValue "parser")
+          , _source   = Just "hlint"
+          , _message  = T.unlines [T.pack msg,T.pack contents]
+          , _relatedInformation = Nothing
+          , _tags     = Nothing
+        }
+      -- This one is defined in Development.IDE.GHC.Error but here
+      -- the types could come from ghc-lib or ghc
+      srcSpanToRange :: SrcSpan -> LSP.Range
+      srcSpanToRange (RealSrcSpan span) = Range {
+          _start = LSP.Position {
+                _line = srcSpanStartLine span - 1
+              , _character  = srcSpanStartCol span - 1}
+        , _end   = LSP.Position {
+                _line = srcSpanEndLine span - 1
+             , _character = srcSpanEndCol span - 1}
+        }
+      srcSpanToRange (UnhelpfulSpan _) = noRange
+
+getIdeas :: NormalizedFilePath -> Action (Either ParseError [Idea])
+getIdeas nfp = do
+  logm $ "getIdeas:file:" ++ show nfp
+  (flags, classify, hint) <- useNoFile_ GetHlintSettings
+
+  let applyHints' (Just (Right modEx)) = Right $ applyHints classify hint [modEx]
+      applyHints' (Just (Left err)) = Left err
+      applyHints' Nothing = Right []
+
+  fmap applyHints' (moduleEx flags)
+
+  where moduleEx :: ParseFlags -> Action (Maybe (Either ParseError ModuleEx))
+#ifndef GHC_LIB
+        moduleEx _flags = do
+          mbpm <- getParsedModule nfp
+          case mbpm of
+            Nothing -> return Nothing
+            Just pm -> do
+              let anns = pm_annotations pm
+              let modu = pm_parsed_source pm
+              return $ Just $ Right (createModuleEx anns modu)
+#else
+        moduleEx flags = do
+          flags' <- setExtensions flags
+          Just <$> (liftIO $ parseModuleEx flags' (fromNormalizedFilePath nfp) Nothing)
+
+        setExtensions flags = do
+          hsc <- hscEnv <$> use_ GhcSession nfp
+          let dflags = hsc_dflags hsc
+          let hscExts = EnumSet.toList (extensionFlags dflags)
+          logm $ "getIdeas:setExtensions:hscExtensions:" ++ show hscExts
+          let hlintExts = mapMaybe (GhclibParserEx.readExtension . show) hscExts
```

Comment:
Having to cast the types from the "real" ghc used by ghcide and the ghc-lib used by hlint is not good at all.
Not sure if there is an alternative.